### PR TITLE
VZ-2108: Make all logging scope fields optional

### PR DESCRIPTION
**Pull request by fredtibbitts**
# Description

All logging scope fields are now optional, except for the workload refs field, because the OAM operator cannot handle a logging scope missing workload refs. The Fluentd image now comes from an environment variable set on the Verrazzano application operator, which is configured by the installer. The Elasticsearch URL and the secret name defaults are still hard-coded (that may change in the future).

Notes:

- I moved the function to fetch the logging scope given labels out of the workload functions file and into the logging scope file. The function has grown in scope, beyond just a simple workload utility function. As a result, the linter complained that the name should drop "LoggingScope" from the front to prevent stuttering when prefixed with the package.
- The defaults apply to multi-cluster logging scopes as well, so now instead of setting MC logging properties on the default logging scope, they can apply to user-defined logging scopes.

# Checklist 

As the author of this PR, I have:

- [x] Checked that I included or updated copyright and license notices in all files that I altered
- [x] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: application-operator/controllers/webhooks/logging_scope_defaulter.go [hunk 1]
```diff
@@ -121,12 +110,9 @@ func ensureDefaultLoggingScope(c client.Client, appConfig *oamv1.ApplicationConf
 		var scope *vzapi.LoggingScope
 		scope, err = fetchLoggingScope(context.TODO(), c, namespacedName)
 		if scope == nil && err == nil {
-			// We might be running in a managed cluster - fetch the Elasticsearch Details to use in
-			// that case
-			elasticSearchDetails := clusters.FetchManagedClusterElasticSearchDetails(context.TODO(), c, log)
```
Comment (desagar) [inline application-operator/controllers/webhooks/logging_scope_defaulter.go:126]: why was this removed? This is the code that determines whether to send logs to the managed cluster elasticsearch.

Comment (desagar) [reply application-operator/controllers/webhooks/logging_scope_defaulter.go:126]: oh wait, never mind. Let me look at the applyDefaults a bit more closely in the controller

File: application-operator/controllers/loggingscope/loggingscope_controller.go [hunk 2]
```diff
@@ -208,3 +211,54 @@ func toResource(workload *unstructured.Unstructured) vzapi.QualifiedResourceRela
...
+	return nil, nil
+}
+
+// applyDefaults fills in any empty fields in the logging scope - also handle the case
+// where we are running in a managed cluster
+func applyDefaults(cli client.Reader, scope *vzapi.LoggingScope) {
+	if scope.Spec.ElasticSearchURL == "" && scope.Spec.SecretName == "" {
+		// if we're running in a managed cluster, use the multicluster ES URL and secret, and if we're
+		// not the fields will be empty and we will set these fields to defaults below
+		elasticSearchDetails := clusters.FetchManagedClusterElasticSearchDetails(context.TODO(), cli)
```
Comment (desagar) [inline application-operator/controllers/loggingscope/loggingscope_controller.go:250]: elasticsearchDetails should only be used on an "all-or-nothing" basis. That function (FetchManagedClusterElasticsearchDetails) should return an empty ES details if any one of the fields is not populated, but it would be good to make sure of that here as well (if elasticsearchDetails.URL and secret are BOTH populated, use it, else ignore it)

File: application-operator/controllers/webhooks/logging_scope_defaulter.go [hunk 3]
```diff
@@ -64,27 +61,19 @@ func (d *LoggingScopeDefaulter) Cleanup(appConfig *oamv1.ApplicationConfiguratio
...
-func CreateDefaultLoggingScope(name types.NamespacedName, esDetails clusters.ElasticsearchDetails) *vzapi.LoggingScope {
-	// If Elasticsearch are provided, use them. Otherwise use the defaults.
-	esURL := defaultElasticSearchURL
-	esSecret := defaultSecretName
-	if esDetails.URL != "" && esDetails.SecretName != "" {
-		esURL = esDetails.URL
-		esSecret = esDetails.SecretName
-	}
-	return &vzapi.LoggingScope{
+func createDefaultLoggingScope(name types.NamespacedName) *vzapi.LoggingScope {
```
Comment (desagar) [inline application-operator/controllers/webhooks/logging_scope_defaulter.go:64]: If no logging scope is specified, and it's a managed cluster, I'm not sure if this will still attach the right logging scope. Has this case been tested?

File: application-operator/controllers/webhooks/logging_scope_defaulter_test.go [hunk 4]
```diff
@@ -229,69 +219,6 @@ func TestLoggingScopeDefaulter_Cleanup(t *testing.T) {
 	mocker.Finish()
 }
 
-// TestCreateDefaultLoggingScope tests the behavior of CreateDefaultLoggingScope
```
Comment (desagar) [inline application-operator/controllers/webhooks/logging_scope_defaulter_test.go:232]: Why were these 2 tests removed? I added one of them to test the managed cluster case and the other to provide a more "pure" unit test for createDefaultLoggingScope.

File: application-operator/controllers/webhooks/logging_scope_defaulter.go (hunk 3 above)
Comment (desagar) [reply application-operator/controllers/webhooks/logging_scope_defaulter.go:64]: looks like that's one of the tests removed from the logging scope defaulter test - let's discuss this to make sure those cases are still being tested.

File: application-operator/controllers/webhooks/logging_scope_defaulter.go [hunk 5]
```diff
@@ -64,27 +61,19 @@ func (d *LoggingScopeDefaulter) Cleanup(appConfig *oamv1.ApplicationConfiguratio
 }
 
 // CreateDefaultLoggingScope creates the default logging scope for the given namespace
```
Comment (kminder) [inline application-operator/controllers/webhooks/logging_scope_defaulter.go:63]: Fix function name in comment to match case.

File: examples/bobs-books/bobs-books-logging-scope.yaml [hunk 6]
```diff
@@ -7,7 +7,6 @@ metadata:
   name: logging-scope
   namespace: bobs-books
 spec:
-  fluentdImage: ghcr.io/verrazzano/fluentd-kubernetes-daemonset:v1.10.4-20201016214205-7f37ac6
```
Comment (kminder) [inline examples/bobs-books/bobs-books-logging-scope.yaml:10]: I'm not even sure this file should be included in this example.

File: application-operator/controllers/loggingscope/loggingscope_controller.go [hunk 7]
```diff
@@ -208,3 +211,54 @@ func toResource(workload *unstructured.Unstructured) vzapi.QualifiedResourceRela
 		Kind:       workload.GetKind(),
 	}
 }
+
+// FromWorkloadLabels returns the LoggingScope object associated with the workload or nil if
+// there is no associated logging scope. The workload lookup is done using the OAM labels from the workload metadata.
+func FromWorkloadLabels(ctx context.Context, cli client.Reader, namespace string, labels map[string]string) (*vzapi.LoggingScope, error) {
```
Comment (kminder) [inline application-operator/controllers/loggingscope/loggingscope_controller.go:217]: Should this be NewFromWorkLoadLabels and if so should it be NewLoggingScopeFromWorkloadLabels because the "New" prefix prevents the repetition with the package name?

File: application-operator/controllers/webhooks/logging_scope_defaulter.go (hunk 3 above)
Comment (fredtibbitts) [reply application-operator/controllers/webhooks/logging_scope_defaulter.go:64]: If no logging scope is specified, then the default logging scope with an empty spec will be attached. When the logging scope is actually used, we check to see if this is a managed cluster and populate the logging scope fields with the MC URL and secret.

File: application-operator/controllers/webhooks/logging_scope_defaulter_test.go (hunk 4 above)
Comment (fredtibbitts) [reply application-operator/controllers/webhooks/logging_scope_defaulter_test.go:232]: I removed these and added corresponding tests in the logging scope controller test file. I did check the code coverage to make sure all of the MC logging scope code was still tested, but it's certainly possible that I missed cases. We can discuss in the zoom review.

**Comment by kminder** [comment]
I think application-operator/deploy/verrazzano.yaml_template should also be updated to keep it in sync until such time as it is deleted.

**Comment by kminder** [comment]
Doesn't application-operator/config/crd/bases/clusters.verrazzano.io_multiclusterloggingscopes.yaml need to be replicated to the platform operator copy of the helm chart also?

File: application-operator/controllers/loggingscope/loggingscope_controller.go (hunk 7 above)
Comment (fredtibbitts) [reply application-operator/controllers/loggingscope/loggingscope_controller.go:217]: It's not creating a logging scope, it's fetching one, or is "New" idiomatic go in this situation? I'll rename it if it's the right thing to do.

File: examples/bobs-books/bobs-books-logging-scope.yaml (hunk 6 above)
Comment (fredtibbitts) [reply examples/bobs-books/bobs-books-logging-scope.yaml:10]: Yeah the examples use a mix of default and user-supplied logging scopes, but the user-supplied scopes just specify the defaults. It seemed like a good idea to have an example that specifies a logging scope with at least one field that gets defaulted. We can remove this logging scope but other examples do the same thing.

**Comment by fredtibbitts** [comment]
> Doesn't application-operator/config/crd/bases/clusters.verrazzano.io_multiclusterloggingscopes.yaml need to be replicated to the platform operator copy of the helm chart also?

@kminder yes, it is replicated, I see it in the files changed list.

File: application-operator/controllers/webhooks/logging_scope_defaulter.go (hunk 5 above)
Comment (fredtibbitts) [reply application-operator/controllers/webhooks/logging_scope_defaulter.go]: Fixed.

**Comment by fredtibbitts** [comment]
> I think application-operator/deploy/verrazzano.yaml_template should also be updated to keep it in sync until such time as it is deleted.

@kminder done. That YAML template is out-of-date, what is it actually used for?

File: application-operator/controllers/loggingscope/loggingscope_controller.go (hunk 2 above)
Comment (fredtibbitts) [reply application-operator/controllers/loggingscope/loggingscope_controller.go:250]: Added a check and it now logs the condition if the secret name is set but the Elasticsearch URL is empty.

File: application-operator/controllers/loggingscope/loggingscope_controller.go (hunk 7 above)
Comment (fredtibbitts) [reply application-operator/controllers/loggingscope/loggingscope_controller.go]: Changed the name to be more descriptive.

File: application-operator/controllers/loggingscope/loggingscope_controller.go [hunk 8]
```diff
@@ -243,13 +243,18 @@ func FromWorkloadLabels(ctx context.Context, cli client.Reader, namespace string
...
 		// not the fields will be empty and we will set these fields to defaults below
 		elasticSearchDetails := clusters.FetchManagedClusterElasticSearchDetails(context.TODO(), cli)
-		scope.Spec.ElasticSearchURL = elasticSearchDetails.URL
-		scope.Spec.SecretName = elasticSearchDetails.SecretName
+		if elasticSearchDetails.URL != "" && elasticSearchDetails.SecretName != "" {
+			scope.Spec.ElasticSearchURL = elasticSearchDetails.URL
+			scope.Spec.SecretName = elasticSearchDetails.SecretName
+		} else if elasticSearchDetails.SecretName != "" && elasticSearchDetails.URL == "" {
+			// Elasticsearch secret is misconfigured, log it and set defaults below
+			log.Info("Unable to set defaults in logging scope for managed cluster because URL is empty")
```
Comment (desagar) [inline application-operator/controllers/loggingscope/loggingscope_controller.go:256]: I'd personally prefer a log level of warning here

**Review by desagar — APPROVED** [review]

File: application-operator/controllers/loggingscope/loggingscope_controller.go (hunk 8 above)
Comment (fredtibbitts) [reply application-operator/controllers/loggingscope/loggingscope_controller.go:256]: This logger only has `Error` and `Info`. I think we have a backlog ticket to change the logger, not sure if that one offers more granularity.
